### DISCUSSION TITLE
Can not get balance

### DISCUSSION THREAD

**Discussion by vodanh86** [ton-org/game-engines-sdk#7] [General]
I use example from this repo: https://github.com/ton-community/flappy-bird
When I start the project, I got this error message: ui.ts:162 failed to load balance 
from these lines of code:
`    async getBalance() {
        try {
            const jetton = this.gameFi.openJetton(Address.parse(this.config.TOKEN_MASTER));
            const jettonWallet = await jetton.getWallet(Address.parse(this.gameFi.wallet.account.address));
            const jettonWalletData = await jettonWallet.getData();
            return jettonWalletData.balance;
        } catch (e) {
            console.error('failed to load balance', e);
            return BigInt(0);
        }
    }`
Do you know how to fix?

**Comment by barinbritva** [comment]
Hi @vodanh86,

Could you provide some details? Did you set up your .env with wallets and everything? Did you try to go through step by step tutorial from the related article? https://docs.ton.org/develop/dapps/tutorials/building-web3-game

**Comment by vodanh86** [comment]
Yes, I did follow every step. This is the .env file:

NODE_ENV=development

# Client Configuration
API_URL=http://127.0.0.1:4000/api
MINI_APP_URL=https://t.me/mybot/a

# Web Server Configuration
CORS_ENABLED=true
CORS_ORIGIN=https://t.me/mybot/a
NGROK_ENABLED=false
NGROK_AUTHTOKEN=
NGROK_DOMAIN=t.me/mybot/a

# Network Configuration
NETWORK=testnet

# Your MNEMONIC
MNEMONIC="*****"

# Pinata API Keys
PINATA_API_KEY=*****
PINATA_SECRET=******

# Telegram Bot Token
TELEGRAM_BOT_TOKEN=******
# Jetton
JETTON_ADDRESS="0QBI7qSmcLXLBzIe00pvYquviANpbpeZ-b9oB5XYZfbA0_ac"

# SBT Collection
FIRST_TIME_SBT_COLLECTION_ADDRESS="0QAmhsTbd5e6lLEOLkHbEjkjFbz7Kek3JyS2RIqrS4IfFrwP"
FIVE_TIMES_SBT_COLLECTION_ADDRESS="0QBi0nm6Z4EZkfuXdSKbhN_P-cbrk8AC-9mqDt2yylwgJjDs"

**Comment by barinbritva** [comment]
Hi @vodanh86,

It seems the connected wallet has no `0QBI7qSmcLXLBzIe00pvYquviANpbpeZ-b9oB5XYZfbA0_ac` kind of jettons. Knowing nothing about this jettons wallet can't tell the balance. It means there is 0 jettons. We put 0 balance in `catch` case. So, it's OK case for the current example. Probably in real word app it can be handled more pretty

For the example: I connected 2 wallets. One of them has Flappy coins, the second one has not. Wallet which has jettons shows correct balance.
`https://tonscan.org/address/UQCxHX_w7PUX2DqeCitOC9DSg1Orpte-qfGO0jM9F3DZYN1H#tokens` - no jettons -> error while loading balance - show 0 balans
`https://tonscan.org/address/UQBH6P5-1KqUNGawyA0cRRqhsD-aZiIsEyr2mycIRmTW406D#tokens` - has 5 jettons -> show correct balance

Try to play a couple of times to earn Flappy coins and you should see the balance will shown as expected

I hope it helps

**Reply by vodanh86** [reply]
Dear Barinbritva,
Thank you for you help. In addition, when I run the project, the TonConnec of GameFi opens the wallet in Telegram. The project is using Testnet, but I can not change the Wallet in Telegram to Testnet.
 Do you know how to change TonConnect to open TonKeeper, so I can use testnet to test the game?

**Reply by barinbritva** [reply]
Nice!

To connect different wallet, do the following: open `workspaces\client\src\connect-wallet-ui.ts` and add `walletApp: 'tonkeeper'` param to `ConnectWalletCanvasScene` constructor:
 
```typescript
return new ConnectWalletCanvasScene(
    gameFi,
    {
        style: 'light',
        onError: (error) => {
            console.error('Caught Error', error);
        },
        // here 👇
        walletApp: 'tonkeeper'
    }
);
```

**Reply by vodanh86** [reply]
Dear Barinbritva,
Thank you so much, I can connect using TestNet in Tonkeeper. However, I still can not get the balance. client calls the "/played" endpoint successfully. 
<img width="1083" alt="Screenshot 2024-07-10 at 17 07 34" src="https://github.com/ton-org/game-engines-sdk/assets/2775170/2c0c509f-9c64-4328-b380-eb32e6189a68">
But the "get_wallet_data" still can not get the result. 

<img width="1071" alt="Screenshot 2024-07-10 at 17 09 13" src="https://github.com/ton-org/game-engines-sdk/assets/2775170/0cc5dd33-9548-4ea2-991c-d0e3d568c68d">

**Reply by barinbritva** [reply]
Alright, one step closer!

I've noticed in `/played` request there is empty `tg_data` which shouldn't be empty. I've seen such bugs related to Telegram clients (MacOS especially). It can be fixed just by closing the Telegram and launch it again. To double check it, after Telegram relaunch, run `window.Telegram.WebApp.initData` in the console

After you see `tg_data` is not empty, the request should work fine. If it doesn't help please send `Response` tab of `/played` endpoint to me be able to see an exact error

**Reply by vodanh86** [reply]
Dear Barin,
I am runing this project from browser directly, so the tg_data is empty. This is the response:
<img width="943" alt="Screenshot 2024-07-11 at 14 09 29" src="https://github.com/ton-org/game-engines-sdk/assets/2775170/fe4591be-b2bc-4b89-98e7-66bbdf1f821c">
Is the project need to run in Telegram?

**Reply by barinbritva** [reply]
Right, this example is designed to showcase TON integration to Mini Apps. The `tg_data` serves to implement seamless authentication. It allows to get user ID and create internal account inside of the game.

If you want to try running in browser I believe you can do it. In `workspaces\server\src\index.ts` replace `const telegramData = processTelegramData(req.tg_data, config.TELEGRAM_BOT_TOKEN!);` with some stub data, like hardcode it. `const telegramData = [YOUR HARDCODED data]`

As I can see it's only blocker to run the example outside of Mini App platform

**Reply by vodanh86** [reply]
Hi Barinbritva,
Finally, I can run this project. This project is so exciting
Thank you so much.

**Reply by vodanh86** [reply]
One more thing, 
Do you know why the Jetton and Sbt tokens are unverified? 

![Screenshot 2024-07-11 at 14 36 08](https://github.com/ton-org/game-engines-sdk/assets/2775170/a50b23ab-6a43-4b64-9b2a-1c2c9be088f9)

**Reply by barinbritva** [reply]
Tonkeeper and some other wallets try to protect users from possible scam. So, they don't have information about tokens and warn users about that

If you run some real token, I believe, there is a procedure to get verified status. But I'm not familiar with the process

**Comment by vodanh86** [comment]
Many thanks, Barinbritva !!!

**Reply by barinbritva** [reply]
You're welcome!
